### ISSUE TITLE
DataModules train on validation data on MultiGPU

### ISSUE THREAD

**Issue by jbuckman** [PyTorchLightning/pytorch-lightning#7728]
## 🐛 Bug

The current codebase makes it extremely easy to make a basic scientific error: training on validation sets. When random operations are used in the `setup()` of a DataModule, this randomness is not replicated among nodes, leading to the construction of different training sets on each node. When train/val sets are constructed via splitting, this means that some data in the training set of one node, may be in the validation set of another; thus, overall, the model will train on the validation set.

This can be seen in the MNIST example [here](https://pytorch-lightning.readthedocs.io/en/latest/extensions/datamodules.html). When `random_split()` is called in `setup()` on MultiGPU, it assigns a different random split to each GPU, leading to overlap between the validation and train sets.

### Expected behavior

The random splits persist across nodes, leading to a well-defined train/val split.

**Comment by awaelchli** [comment]
Hi you can either set a seed globally for all nodes using 

```python
from pytorch_lightning import seed_everything
seed_everything(123)
```

or have `random_split()` take a [specific seed](https://pytorch.org/docs/stable/data.html#torch.utils.data.random_split) for just the splitting.

**Comment by jbuckman** [comment]
Awesome, very convenient! I would suggest including it in the DataModule example code, as this mistake seems easy to make, and causes experiments to silently give the wrong outcomes. Right now, it is a key place where Lightning's philosophy of "write it for 1 machine and it seamlessly works on N machines" breaks down.
